weavertest/internal/chain/chain_test.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chain
+
+import (
+	"context"
+	"testing"
+
+	"github.com/ServiceWeaver/weaver/weavertest"
+)
+
+func TestOneComponentImpl(t *testing.T) {
+	// Tests weaver.Test with a component implementation pointer argument.
+	for _, runner := range weavertest.AllRunners() {
+		runner.Test(t, func(t *testing.T, a A, d *d) {
+			if err := a.Propagate(context.Background(), 1); err != nil {
+				t.Fatal(err)
+			}
+			if got, want := d.val, 4; got != want {
+				t.Fatalf("got %d, want %d", got, want)
+			}
+		})
+	}
+}
+
+func BenchOneComponentImpl(b *testing.B) {
+	// Tests weaver.Bench with a component implementation pointer argument.
+	for _, runner := range weavertest.AllRunners() {
+		runner.Bench(b, func(b *testing.B, a A, d *d) {
+			if err := a.Propagate(context.Background(), 1); err != nil {
+				b.Fatal(err)
+			}
+			if got, want := d.val, 4; got != want {
+				b.Fatalf("got %d, want %d", got, want)
+			}
+		})
+	}
+}
+
+func TestTwoComponentImpls(t *testing.T) {
+	// Tests weaver.Test with two component implementation pointer arguments.
+	for _, runner := range weavertest.AllRunners() {
+		runner.Test(t, func(t *testing.T, a A, c *c, d *d) {
+			if err := a.Propagate(context.Background(), 1); err != nil {
+				t.Fatal(err)
+			}
+			if got, want := c.val, 3; got != want {
+				t.Fatalf("got %d, want %d", got, want)
+			}
+			if got, want := d.val, 4; got != want {
+				t.Fatalf("got %d, want %d", got, want)
+			}
+		})
+	}
+}
+
+func TestThreeComponentImpls(t *testing.T) {
+	// Tests weaver.Test with three component implementation pointer arguments.
```

Comment:
What code are we covering here that wasn't covered by the two component test?

Reply:
None. I removed `D` completely and simplified things to a chain of `A`, `B`, and `C`.